### PR TITLE
Managed length in _sh_read_data_raw_get

### PR DESCRIPTION
**Pull request by albezanc**
I need to connect two iis2iclx using the "sensor hub" functionality in this configuration:

STM32XXXX -> SPI -> SENSOR- > I2C (SensorHub) -> SENSOR

By doing this I started with another example of "Sensor Hub".  [lsm6dsox_sh_fifo_lps22hh.c](https://github.com/Tecnosoft/STMems_Standard_C_drivers/blob/master/lsm6dsox_STdC/examples/lsm6dsox_sh_fifo_lps22hh.c)  

By adapting the "_read function_", on the sensor hub side from the example  above it should be as follow:

```
static int32_t iis2iclx_read_iis2iclx_cx(void *ctx, uint8_t reg,
                                        uint8_t *data, uint16_t len)
{
  iis2iclx_sh_cfg_read_t sh_cfg_read;
  int16_t data_raw_acceleration[3];
  int32_t ret;
  uint8_t drdy;
  iis2iclx_status_master_t master_status;
  /* Disable accelerometer. */
  iis2iclx_xl_data_rate_set(&ag_ctx, IIS2ICLX_XL_ODR_OFF);
  /* Configure Sensor Hub to read LPS22HH. */
  sh_cfg_read.slv_add = (IIS2ICLX_I2C_ADD_L & 0xFEU) >> 1; /* 7bit I2C address */
  sh_cfg_read.slv_subadd = reg;
  sh_cfg_read.slv_len = len;
  ret = iis2iclx_sh_slv0_cfg_read(&ag_ctx, &sh_cfg_read);
  iis2iclx_sh_slave_connected_set(&ag_ctx, IIS2ICLX_SLV_0);
  /* Enable I2C Master and I2C master. */
  iis2iclx_sh_master_set(&ag_ctx, PROPERTY_ENABLE);
  /* Enable accelerometer to trigger Sensor Hub operation. */
  iis2iclx_xl_data_rate_set(&ag_ctx, IIS2ICLX_XL_ODR_104Hz);
  /* Wait Sensor Hub operation flag set. */
  iis2iclx_acceleration_raw_get(&ag_ctx, data_raw_acceleration);

  do {
    HAL_Delay(20);
    iis2iclx_xl_flag_data_ready_get(&ag_ctx, &drdy);
  } while (!drdy);

  do {
    //HAL_Delay(20);
    iis2iclx_sh_status_get(&ag_ctx, &master_status);
  } while (!master_status.sens_hub_endop);

  /* Disable I2C master and XL(trigger). */
  iis2iclx_sh_master_set(&ag_ctx, PROPERTY_DISABLE);
  iis2iclx_xl_data_rate_set(&ag_ctx, IIS2ICLX_XL_ODR_OFF);
  /* Read SensorHub registers. */
  iis2iclx_sh_read_data_raw_get(&ag_ctx, (iis2iclx_emb_sh_read_t *)data);
  return ret;
}
````

This function is typically used by others API (ie.` _device_id_get`) and the input argument `uint8_t *data` is typically an array of size `uint16_t len` that is (frequently < 18) but the call 
` iis2iclx_sh_read_data_raw_get(&ag_ctx, (iis2iclx_emb_sh_read_t *)data);` always overwrite 18 bytes.

Therefore the patch adds a check in `iis2iclx_sh_read_data_raw_get` to avoid writing outside the bounds of the `uint8_t *data` array.

### CODE REVIEW

File: iis2iclx_reg.c [hunk 1]
```diff
@@ -6114,21 +6114,23 @@ int32_t iis2iclx_mlc_out_get(const stmdev_ctx_t *ctx, uint8_t *buff)
   *
   */
 int32_t iis2iclx_sh_read_data_raw_get(const stmdev_ctx_t *ctx,
-                                      iis2iclx_emb_sh_read_t *val)
+                                      iis2iclx_emb_sh_read_t *val, uint16_t len)
 {
   int32_t ret;
 
-  ret = iis2iclx_mem_bank_set(ctx, IIS2ICLX_SENSOR_HUB_BANK);
+  if (len < 19) {
```
Comment (avisconti) [inline iis2iclx_reg.c]: I prefer that you exit here. If you want to be MISRA compliant and have only one point of exit then just do:

   ```
   if (len > 18)
   {
      ret = -1; (or -EINVAL)
      got exit;
   }
```

And pls add also a comment before the test ;)

File: iis2iclx_reg.c [hunk 2]
```diff
@@ -6835,10 +6837,11 @@ int32_t iis2iclx_sh_slv0_cfg_read(const stmdev_ctx_t *ctx,
 
   if (ret == 0)
   {
-    slv0_add.slave0 = (uint8_t) val->slv_add >> 1;
+    slv0_add.slave0 = (uint8_t) val->slv_add;
```
Comment (avisconti) [inline iis2iclx_reg.c]: good catch, but put it in a different commit with proper commit message

**Review by avisconti — CHANGES_REQUESTED** [review]
Good catch, thanks

File: iis2iclx_reg.c (hunk 1 above)
Comment (albezanc) [reply iis2iclx_reg.c]: It should be ok now

File: iis2iclx_reg.c (hunk 2 above)
Comment (albezanc) [reply iis2iclx_reg.c]: Done

File: iis2iclx_reg.c [hunk 3]
```diff
@@ -6114,23 +6114,30 @@ int32_t iis2iclx_mlc_out_get(const stmdev_ctx_t *ctx, uint8_t *buff)
...
 int32_t iis2iclx_sh_read_data_raw_get(const stmdev_ctx_t *ctx,
-                                      iis2iclx_emb_sh_read_t *val)
+                                      iis2iclx_emb_sh_read_t *val, uint16_t len)
 {
   int32_t ret;
 
+  /* Check on registers SENSOR_HUB_X range */
+  if (len > 18)
+  {
+    ret = -1; /* (or -EINVAL) */
```
Comment (avisconti) [inline iis2iclx_reg.c]: I was suggesting either one or the other. Maybe ```ret = -EINVAL;``` is better? What do you think? Otherwise leave ```ret = -1;```

Comment (avisconti) [reply iis2iclx_reg.c]: rest is good!!

File: iis2iclx_reg.c (hunk 1 above)
Comment (avisconti) [reply iis2iclx_reg.c]: @albezanc it is almost perfect. Only one note on the ```ret = -1;``` See below

File: iis2iclx_reg.c (hunk 3 above)
Comment (albezanc) [reply iis2iclx_reg.c]: Honestly, I prefer  `-EINVAL` but it needs to include `errno.h` that is actually a standard but unfortunately it is not included in these drivers. 

I don't like to introduce the error code `-1` because it is undefined and can be in contrast with other error definition but without `errno.h` I can only suggest to use ` -EINVAL` in a comment.

Comment (avisconti) [reply iis2iclx_reg.c]: OK, I still have some doubts, but it may make sense. Anyway I would at least remove the word "or", i.e. like this:

```ret = -1; /* -EINVAL */```

I don't like he word "or" because it seems a user choice which is not. It is more an indication of the meaning.
Ah! Another possibility would be to use:

```ret = -22; /* -EINVAL */```

because it should be a standard value.

Comment (avisconti) [reply iis2iclx_reg.c]: I would use the former because it is simpler.

Comment (albezanc) [reply iis2iclx_reg.c]: I like this one: `ret = -22; /* -EINVAL */` 👍

**Review by avisconti — APPROVED** [review]
Thanks!
